File: 51-n-queens/n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def isSafe(board, row, col):
            # Check upward in the same column
            for i in range(row):
                if board[i][col] == 'Q':
                    return False

            # Check upper-left diagonal
            i, j = row - 1, col - 1
            while i >= 0 and j >= 0:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j -= 1

            # Check upper-right diagonal
            i, j = row - 1, col + 1
            while i >= 0 and j < n:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j += 1

            return True

        def solve(row):
            if row == n:
                res.append(["".join(r) for r in board])
                return

            for col in range(n):
                if isSafe(board, row, col):
                    board[row][col] = 'Q'
                    solve(row + 1)
                    board[row][col] = '.'  # backtrack

        board = [["." for _ in range(n)] for _ in range(n)]
        res = []
        solve(0)
        return res

File: 51-n-queens/n-queens.cpp
class Solution {
public:
    bool isSafe(vector<string>& board, int row, int col) {
        int N = board.size();
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') return false;
        }

        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j]== 'Q') return false;
        }

        for (int i = row, j = col; i >= 0 && j < N; i--, j++) {
            if (board[i][j]== 'Q') return false;
        }

        return true;
    }


    void solver(vector<vector<string>>& result ,vector<string>& curr,int row = 0){
        int N = curr.size();

        if (row == N){
            result.push_back(curr);
            return;
        }

        for (int col = 0 ; col < N ; col++){
            if (isSafe(curr,row,col)){
                curr[row][col] = 'Q';
                solver(result,curr,row+1);
                curr[row][col] = '.';
            }
        }

    }
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> result;
        vector<string> curr(n , string(n,'.'));

        solver(result,curr);

        return result;
    }
};